efibootmanager_policy.go

```diff
@@ -0,0 +1,335 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"sort"
+
+	"github.com/canonical/go-tpm2"
+	"github.com/chrisccoulson/tcglog-parser"
+	"github.com/snapcore/secboot/internal/pe1.14"
+
+	"golang.org/x/xerrors"
+)
+
+// computePeImageDigest computes a hash of a PE image in accordance with the "Windows Authenticode Portable Executable Signature
+// Format" specification. This function interprets the byte stream of the raw headers in some places, the layout of which are
+// defined in the "PE Format" specification (https://docs.microsoft.com/en-us/windows/win32/debug/pe-format)
+func computePeImageDigest(alg tpm2.HashAlgorithmId, image EFIImage) (tpm2.Digest, error) {
+	r, err := image.Open()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot open image: %w", err)
+	}
+	defer r.Close()
+
+	var dosheader [96]byte
+	if _, err := r.ReadAt(dosheader[0:], 0); err != nil {
+		return nil, err
+	}
+
+	var coffHeaderOffset int64
+	if dosheader[0] == 'M' && dosheader[1] == 'Z' {
+		signoff := int64(binary.LittleEndian.Uint32(dosheader[0x3c:]))
+		var sign [4]byte
+		r.ReadAt(sign[:], signoff)
+		if !(sign[0] == 'P' && sign[1] == 'E' && sign[2] == 0 && sign[3] == 0) {
+			return nil, fmt.Errorf("invalid PE COFF file signature: %v", sign)
+		}
+		coffHeaderOffset = signoff + 4
+	}
+
+	p, err := pe.NewFile(r)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot decode PE binary: %w", err)
+	}
+
+	var isPe32Plus bool
+	var sizeOfHeaders int64
+	var dd []pe.DataDirectory
+	switch oh := p.OptionalHeader.(type) {
+	case *pe.OptionalHeader32:
+		sizeOfHeaders = int64(oh.SizeOfHeaders)
+		dd = oh.DataDirectory[0:oh.NumberOfRvaAndSizes]
+	case *pe.OptionalHeader64:
+		isPe32Plus = true
+		sizeOfHeaders = int64(oh.SizeOfHeaders)
+		dd = oh.DataDirectory[0:oh.NumberOfRvaAndSizes]
+	default:
+		return nil, errors.New("PE binary doesn't contain an optional header")
+	}
+
+	// 1) Load the image header in to memory.
+	hr := io.NewSectionReader(r, 0, sizeOfHeaders)
```

Comment:
should we check that sizeOfHeaders is not unreasonably large?

Reply:
I've add a helper to hash data 1 page at a time now

Reply:
thank you